WIP: add SparseMerklePath

Qyriad opened the pull request:
## Describe your changes

This implements a Merkle path representation which compacts empty nodes into a bitmask to save memory, currently as as `u64` — do we need to worry about cases with more than 64 empty nodes in a path?

I used the representation suggested by https://github.com/0xPolygonMiden/crypto/issues/215#issuecomment-1792720322, but without the `depth` field, as to my understanding it can be calculated as the sum of stored nodes and set bits in the bitmask.

I want to confirm this is the desired approach, so I haven't yet implemented most of the helper and conversion methods and traits.

Closes #215.


## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

bobbinth commented on src/merkle/smt/mod.rs:
```diff
@@ -79,28 +79,32 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
-    fn open(&self, key: &Self::Key) -> Self::Opening {
-        let leaf = self.get_leaf(key);
-
+    /// Returns a [`MerklePath`] to the specified key.
```
nit: should this say "Returns a [`MerklePath`] to the leaf containing the specified key."

bobbinth commented on src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,109 @@
+/// [`MerklePath`], for each empty node.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct SparseMerklePath<const DEPTH: u8> {
```
I'm not sure about having `DEPTH` be a generic parameter. Sometimes, we may be dealing with the paths for which we don't know the depth (e.g., paths that come over the wire or are stored in the database).

Qyriad replied:
Note that, to my understanding, we will still need to know the depth at runtime, so that we can determine if a node is empty or not.

bobbinth commented on src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,109 @@
+    pub fn from_path(path: MerklePath) -> Self {
+        // Note that the path does not include the node itself that it is a path to.
+        // That is to say, the path is not inclusive of its ending.
+        assert!(path.len() <= u8::MAX.into(), "SparseMerklePath may have at most 256 items");
```
We should check against `SMT_MAX_DEPTH` here rather than `u8::MAX`. Also, I wonder if we should return an error here instead of panicking.

Qyriad replied:
I made it an assert to match https://github.com/0xPolygonMiden/crypto/blob/aa45474377e978050958958d75688e7a8d46b628/miden-crypto/src/merkle/path.rs#L26

should I change both of them, or just this one?

bobbinth reviewed (commented):
Thank you! Looks good! I left a couple of comments inline. To answer some other questions:

> This implements a Merkle path representation which compacts empty nodes into a bitmask to save memory, currently as as `u64` — do we need to worry about cases with more than 64 empty nodes in a path?

I think think the `u64` assumption is fine as in this library we deal with Sparse Merkle trees which can have max depth of 64 (I left a comment about this inline).

> I used the representation suggested by [#215 (comment)](https://github.com/0xPolygonMiden/crypto/issues/215#issuecomment-1792720322), but without the `depth` field, as to my understanding it can be calculated as the sum of stored nodes and set bits in the bitmask.

I think we need the `depth` field. Without the `depth` field we might have some ambiguity. Using a 4-bit example: `1101` could refer to a path of depth 4, depth 3, or depth 2 - or am i missing something?

bobbinth commented on src/merkle/sparse_path.rs:
I would also implement `Serializable` and `Desirializable` traits for the `SparseMerklePath`.

Qyriad commented:
> Using a 4-bit example: 1101 could refer to a path of depth 4, depth 3, or depth 2 - or am i missing something?

In this PR, the bit-pattern 1101 would indicate that the nodes at depth-0, depth-2, and depth-3 are empty nodes in the path, and thus the full depth of the path is `self.nodes.len() + 3`, as `self.nodes` contains all nodes in the path, except for the 3 empty nodes.

bobbinth commented:
> > Using a 4-bit example: 1101 could refer to a path of depth 4, depth 3, or depth 2 - or am i missing something?
> 
> In this PR, the bit-pattern 1101 would indicate that the nodes at depth-0, depth-2, and depth-3 are empty nodes in the path, and thus the full depth of the path is `self.nodes.len() + 3`, as `self.nodes` contains all nodes in the path, except for the 3 empty nodes.

Makes sense. And how would we encode a path of length 3 where nodes at depth 0 and 2 are empty nodes?

Qyriad commented:
```rust
SparseMerklePath {
    nodes: vec![node_at_depth_1],
    empty_nodes: 0b0101
};
```

Here `nodes.len() == 1`, and `empty_nodes.count_ones() == 2`, so `path.depth() == 3`.

Qyriad commented:
I've implemented the rest of the PR, and addressed your review comments. Note that I can't directly implement things like `From<MerklePath> for SparseMerklePath` and `FromIterator<RpoDigest> for SparseMerklePath`, as those conversion require knowing the tree depth. I also didn't implement `Deref` like `MerklePath` has, since each node has to be calculated for retrieval.

Qyriad commented:
I'm marking this back as a draft to implement `From<SparseMerklePath> for MerklePath` by making the assumption that Merkle paths always go from the root to a leaf, which `SimpleSmt` seems to imply is correct.

@bobbinth if we do need partial paths (like `MerkleTree::get_path()` can return), then I will have to add a field to keep track of the tree's depth.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+/// usage by this struct, but will incur overhead during iteration or conversion to a
+/// [`MerklePath`], for each empty node.
+///
+/// NOTE: This type assumes that Merkle paths always span from the root of the tree to a leaf.
+/// Partial paths are not supported.
```
I think this is probably fine to assume in our context.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+        for (depth, node) in iter::zip(0..tree_depth, iter) {
+            let &equivalent_empty_node = EmptySubtreeRoots::entry(tree_depth, depth);
+            if node == equivalent_empty_node {
+                // FIXME: should we just fallback to the Vec if we're out of bits?
+                assert!(depth < 64, "SparseMerklePath may have at most 64 empty nodes");
```
Given the check on line 34 above, is this really needed?

One thing we can check though is that `SMT_MAX_DEPTH < 64` - but this could be a compile-time check.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+        Self::from_sized_iter(path.into_iter())
+    }
+
+    /// Converts this sparse representation back to a normal [`MerklePath`].
+    pub fn into_path(mut self) -> MerklePath {
```
I would probably get rid of this function and move its code into `impl From<SparseMerklePath> for MerklePath`.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+/// # Panics
+///
+/// This conversion fails and panics if the path length is greater than [`SMT_MAX_DEPTH`].
+impl From<MerklePath> for SparseMerklePath {
```
Should this be a `TryFrom` rather than `From`?

bobbinth reviewed (commented):
Looks good! Thank you! Not a full review but I left some comments inline.

> Note that I can't directly implement things like `From<MerklePath> for SparseMerklePath` and `FromIterator<RpoDigest> for SparseMerklePath`, as those conversion require knowing the tree depth.

I think we can assume that the length of the `MerklePath` is the depth of the tree. Or would this cause issues?

> I'm marking this back as a draft to implement `From<SparseMerklePath> for MerklePath` by making the assumption that Merkle paths always go from the root to a leaf, which `SimpleSmt` seems to imply is correct.

I think we can make this assumption.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+
+/// Owning iterator for [`SparseMerklePath`].
+// TODO: add a non-owning iterator too.
+pub struct SparseMerkleIter {
+    /// The "inner" value we're iterating over.
+    path: SparseMerklePath,
```
```suggestion
pub struct SparseMerkleIter<'path> {
    /// The "inner" value we're iterating over.
    path: &'path SparseMerklePath,
```

Would this work generally? It looks like we don't need mutable or owned access to the merkle path, so could we make this Iter generally hold a reference instead? Then we'd have a non-owning iterator for free.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let next_depth = self.next_depth.unwrap_or(0);
+        let len: usize = self.path.depth().into();
+        let remaining = len - next_depth as usize;
```
Can you add a safety comment to mention why this can never underflow?

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+impl ExactSizeIterator for SparseMerkleIter {
+    fn len(&self) -> usize {
+        let next_depth = self.next_depth.unwrap_or(0);
+        (self.path.depth() - next_depth) as usize
```
Same here regarding the safety comment.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+// SPARSE MERKLE PATH CONTAINERS
```
```suggestion
// SPARSE VALUE PATH
```

Nit: I think this should mention the struct name.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+// SERIALIZATION
+// ================================================================================================
```
I would probably group the serialization for `SparseMerklePath` together with that struct, so maybe before the `SparseMerkleIter`

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+
+/// Owning iterator for [`SparseMerklePath`].
+// TODO: add a non-owning iterator too.
+pub struct SparseMerkleIter {
```
I think just to avoid confusion it would be best to name this `SparseMerklePathIter`.

PhilippGackstatter reviewed (commented):
Looks good!

I think my main comments are:

- It seems it would be best if _by construction_ `SparseMerklePath::depth` would never exceed `SMT_MAX_DEPTH`. Then we can avoid various checked shift operations. It currently seems like we have to handle this potential error in every function, which I think is error prone. If instead we can strengthen the guarantees of the type to validate the above at construction time, I think that would make the coder overall safer and easier to reason about.
- It would be great if we could avoid re-implementing the inner logic of `SparseMerklePath` in `SimpleSmt::get_path` (and probably `Smt::open` once we change it to return a sparse merkle path). Maybe a way to do this could be to implement iterator types that yield the merkle path of a given leaf. The yielded value could be the depth and the associated hash, e.g. something like `(u8, RpoDigest)`. Then `SparseMerklePath::from_path_iter` can take that iterator and check if the yielded value at that depth is an empty one, and set the appropriate bit in the bitmask. That way we would only have to implement such (probably private/internal) iterators for `SimpleSmt` and `Smt`. The bitmask logic of sparse merkle path would then be entirely contained in that type's constructors. That would also avoid the need for `SparseMerklePath::from_raw_parts` which can currently violate the above guarantee.

Let me know what you think of this and if this is possible.

Qyriad commented on miden-crypto/src/merkle/smt/simple/mod.rs:
```diff
@@ -173,6 +174,40 @@ impl<const DEPTH: u8> SimpleSmt<DEPTH> {
 
+    /// Returns a path (but not an opening) to the leaf associated with `key`.
+    ///
+    /// Unlike [`SimpleSmt::open()`], this returns a [SparseValuePath] which has a more efficient
+    /// memory representation optimized for paths containing empty nodes. See [SparseMerklePath]
+    /// for more information.
+    pub fn get_path(&self, key: &LeafIndex<DEPTH>) -> SparseValuePath {
```
That will definitely require some sweeping find-and-replace but at a glance should work out alright. I'll implement `PartialEq<MerklePath> for SparseMerklePath` to facilitate interoperability between the two types.

Qyriad commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+        }
+
+        debug_assert_eq!(sparse_nodes.next(), None);
+        drop(sparse_nodes);
```
Technically it's only required because of the `debug_assert!(self.nodes.is_empty())`, which calls `self.nodes.is_empty()`, which takes a shared reference from `self.nodes`. However if `sparse_nodes` is still alive, then it still holds an exclusive reference to `self.nodes`. Rust isn't quite smart enough to realize that `sparse_nodes` isn't used for the rest of the function and thus can be dropped then.

I should put a comment explaining this, or I can remove the `debug_assert!()`.

Qyriad commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,385 @@
+    /// Get a specific node in this path at a given depth.
+    ///
+    /// # Panics
+    /// With debug assertions enabled, this function panics if `node_depth` is greater than
+    /// `tree_depth` (as it is impossible to have a node of greater depth than the tree it is
+    /// contained in).
```
You're right! That's a holdover from a previous version of that function. Thank you for the catch.

Qyriad commented:
This is ready for review now!

PhilippGackstatter commented on miden-crypto/src/merkle/path.rs:
```diff
@@ -22,6 +28,8 @@ impl MerklePath {
 
     /// Creates a new Merkle path from a list of nodes.
+    ///
+    /// The list is assumed to be in order of deepest to shallowest.
```
```suggestion
    /// The list must be in order of deepest to shallowest.
```

Nit: I would strengthen the wording.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+    }
+}
+
+/// # Errors
+///
+/// This conversion returns [MerkleError::DepthTooBig] if the path length is greater than
+/// [`SMT_MAX_DEPTH`].
+impl TryFrom<MerklePath> for SparseMerklePath {
+    type Error = MerkleError;
+
+    fn try_from(path: MerklePath) -> Result<Self, MerkleError> {
```
Nit: I would add the doc comment at the function level. That way, when I hover over `SparseMerklePath::try_from` in my IDE, I get the docs.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+
+/// Contructs a [SparseMerklePath] out of an iterator of optional nodes, where `None` indicates an
+/// empty node.
+impl FromIterator<(NonZero<u8>, Option<RpoDigest>)> for SparseMerklePath {
```
I don't think we can implement this without allowing users of the type to violate the guarantee that the depth never exceeds SMT_MAX_DEPTH. For example, this test currently passes:

```rust
#[test]
fn test_from_iter() {
    let tree = make_smt(8);

    let (control_path, _) = tree.open(&tree.entries().next().unwrap().0).into_parts();

    let path_depth_iter = path_depth_iter(tree.depth()).into_iter();
    let chain = [(NonZero::new(64u8).unwrap(), Some(control_path[0]))];
    let zipped = path_depth_iter.zip(control_path.into_iter().map(Some));
    let path = SparseMerklePath::from_iter(zipped.chain(chain));

    assert_eq!(path.depth(), 65);
}
```

So I think this should be an inherent, private method only used from the other safe constructors.

gswirski replied:
Agreed that it should be a private method. Since it's only called from `from_iter_with_depth`, I internalized it there and made the whole thing private (as per my other comment).

As for the capacity, we don't know the Vec size until we consume `tree_depth` elements of the iterator and know how many empty elements there are. We don't store empty nodes.

PhilippGackstatter reviewed (commented):
Not a full review yet, just some initial comments.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+// ================================================================================================
+/// A container for a [crate::Word] value and its [SparseMerklePath] opening.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SparseValuePath {
```
I wonder if we should call this `SimpleSmtProof` instead (if this ends up replacing `ValuePath` completely), to be analogous to `SmtProof`.

gswirski replied:
Can make the change now, or we can consider it part of the https://github.com/0xPolygonMiden/crypto/issues/404 ? What would you prefer?

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+    }
+}
+
+/// # Errors
+///
+/// This conversion returns [MerkleError::DepthTooBig] if the path length is greater than
+/// [`SMT_MAX_DEPTH`].
```
Nit: Similar comment as before, I'd move this to function.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+
+impl PartialEq<SparseValuePath> for ValuePath {
+    fn eq(&self, rhs: &SparseValuePath) -> bool {
+        rhs == self
```
```suggestion
        rhs.eq(self)
```

Nit: Would it be better to call the method directly? I'm not quite sure if this would cause infinite recursion right now and when using `eq` the wrong way around, rustc will complain, so this seems safer.

gswirski replied:
The behavior of `==` and `.eq()` is identical so it's mostly a matter of personal preference. If we make the chance, there would be no infinite recursion because we implement `ValuePath == SparseValuePath` using `SparseValuePath == ValuePath` implementation (notice that `rhs` moves to the left-hand side of the operator).

Of course there is a chance that someone would then decide to implement `SparseValuePath == ValuePath` based on this, and then we do get an infinite loop indeed. Luckily, this code is already exercised in tests and would be caught by `miden-crypto/src/merkle/store/tests.rs:193:30`:
```
        assert_eq!(
            store_path.path, smt_path,
            "the returned merkle path does not match the computed values"
        );
```

If you feel strongly about `.eq` I can replace it, but to me the `==` operator feels more natural. There are a total of 4 manual `PartialEq` implementations in this repo, and they all use `==` and `!=` instead of `.eq()`.

PhilippGackstatter replied:
No, makes sense, let's leave it as-is :)

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+            assert_eq!(test_node, control_node);
+        }
+
+        // Rinse and repeat for each remaining depth.
```
This is very explicit which is great on one hand, but on the other, is somewhat tedious to review, so if it's possible to deduplicate this, that'd be great. One idea is two for loops, one which checks all the depths that are empty and the other checks for non-empty nodes.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+            let index = NodeIndex::from(Smt::key_to_leaf_index(key));
+
+            let control_path = tree.get_path(key);
+            for (&control_node, proof_index) in iter::zip(&*control_path, index.proof_indices()) {
+                let proof_node = tree.get_hash(proof_index);
+                assert_eq!(control_node, proof_node, "WHat");
+            }
+
+            let sparse_path =
+                SparseMerklePath::from_sized_iter(control_path.clone().into_iter()).unwrap();
+            for (sparse_node, proof_idx) in iter::zip(sparse_path.clone(), index.proof_indices()) {
+                let proof_node = tree.get_hash(proof_idx);
+                assert_eq!(sparse_node, proof_node, "WHat");
+            }
```
Nit: I'd remove the assert message or make it more useful.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+            // Letting index get to `control_path.len()` will test that both sides correctly return
+            // `None` for out of bounds access.
+            for index in 0..=(control_path.len()) {
+                let control_node = control_path.at_idx(index).copied();
```
Nit: Maybe `MerklePath::at_idx` should return `Option<RpoDigest>`, i.e. do the `copied()` internally?

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+    }
+
+    #[test]
+    fn test_owning_iterator() {
```
I think the name of this test and `test_borrowing_iterator` should be swapped.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+            assert_eq!(control_path.depth(), sparse_path.depth());
+            assert_eq!(sparse_path.depth(), SMT_DEPTH);
+
+            // Test that both iterators yield the same amount of the same values.
```
We should probably drive the iterators manually, because `zip` will stop returning when either one of the iterators returns `None`, so with this we can't make sure that one iterator doesn't actually return more items.

Also applies to the other test.

gswirski replied:
Replaced it with `itertools::zip_eq` to validate length.

Importing `itertools` just for this feels a little bit like using `zeropad`. I added it to dev_dependencies only and it was already a transitive dependency of criterion so maybe it's fine. Let me know if you'd rather have `zip_eq` implemented directly in the project.

PhilippGackstatter replied:
Adding it as dev-dependency seems fine to me.

PhilippGackstatter reviewed (commented):
Looks great! I like the iterator approach and the code is easy to follow.

I think my main comments are around (type) safety overall. Mainly it would be good to clarify whether the intention is to support sparse merkle paths of depth greather than 64 or not. In my opinion 64 is enough for now. In that case though, we should add checks in the iterator constructors and the deserialization code (see inline comments).

gswirski commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+    }
+}
+
+impl<'p> IntoIterator for &'p SparseMerklePath {
```
Hm, I don't see how. `Cow` needs a lifetime specifier which we don't have for the `impl IntoIterator for SparseMerklePath`.

Qyriad replied:
I can experiment with that, though note that std types also duplicate their iterators:

https://github.com/rust-lang/rust/blob/8bf5a8d12feea10dfada53fb2d119283b0e0107c/library/alloc/src/collections/linked_list.rs#L73-L78

https://github.com/rust-lang/rust/blob/8bf5a8d12feea10dfada53fb2d119283b0e0107c/library/alloc/src/collections/linked_list.rs#L141-L146

PhilippGackstatter replied:
My preference would be somewhat on avoiding duplication, so trying the `Cow` approach, but if that doesn't work or results in worse code, then we can leave it as-is.

PhilippGackstatter commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+    /// which nodes are empty nodes.
+    ///
+    /// # Errors
+    /// Returns [MerkleError::DepthTooBig] if `tree_depth` is greater than [SMT_MAX_DEPTH].
```
> creating a series of `n` `SperseMerklePath`s where your iterator contains `n * path_len` elements.

Right, I hadn't thought about this use case, although not sure in what scenario this would actually occur. For now, as you say, the only usage is from within `from_sized_iter`, so the first use case.

Overall, though, making this private is probably the best way to go :+1:.

gswirski commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,813 @@
+/// memory this struct will use. This type calculates empty nodes on-demand when iterated through,
+/// converted to a [MerklePath], or an empty node is retrieved with [`SparseMerklePath::at_idx()`]
+/// or [`SparseMerklePath::at_depth()`], which will incur overhead.
+///
```
64 nodes in total, as per the SMT_MAX_DEPTH. I clarified the comment.

gswirski commented on miden-crypto/src/merkle/sparse_path.rs:
I defer to @bobbinth here. Created a new issue to make sure it's not forgotten: https://github.com/0xPolygonMiden/crypto/issues/404

gswirski commented on miden-crypto/src/merkle/sparse_path.rs:
I think it depends if we fully replace `MerklePath` with `SparseMerklePath`? Included it in https://github.com/0xPolygonMiden/crypto/issues/404

gswirski commented:
Hi @PhilippGackstatter I think I addressed all of the feedback. It's ready for another review. To see only the changes from the last review: https://github.com/0xPolygonMiden/crypto/compare/5ab7980...9daaef8 (but let me know if you want my commits squashed instead).

Like mentioned in inline comments, I created a followup issue for replacing `MerklePath` with `SparseMerklePath` and associated API changes: https://github.com/0xPolygonMiden/crypto/issues/404

PhilippGackstatter reviewed (approved):
Thanks for the updates! Looks good to me!

I like the concise for loop version of the sparse_bits test :slightly_smiling_face:.

The only outstanding thing from my side is the question if we can get rid of the `IntoIter` struct. I left a response to an earlier comment about this.

gswirski commented:
Smart idea with the `Cow`, thank you for explaining this. I've got rid of the `IntoIter` struct.

@bobbinth there is the remaining question of the scope of this PR, otherwise this should be ready to merge.
> I'm approving this with the assumption that the scope of the PR was to only add SparseMerklePath, but not also APIs that use it on the SMTs.

Qyriad commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,692 @@
+        I: IntoIterator<IntoIter: ExactSizeIterator, Item = RpoDigest>,
+    {
+        let iterator = iterator.into_iter();
+        // `iterator.len() as u8` will truncate, but not below `SMT_MAX_DEPTH`, which
+        // `from_iter_with_depth` checks for.
+        Self::from_iter_with_depth(iterator.len() as u8, iterator)
```
That's not strictly necessary. If it doesn't fit into a `u8`, then it will truncate to `u8::MAX`, aka 255. `from_iter_with-depth()` then checks that against `SMT_MAX_DEPTH`, aka 64:
```rust
if tree_depth > SMT_MAX_DEPTH {
    return Err(MerkleError::DepthTooBig(tree_depth as u64));
}
```

So it will correctly error if `iterator.len() as u8` overflows.

bobbinth replied:
> If it doesn't fit into a `u8`, then it will truncate to `u8::MAX`, aka 255.

I'm not sure that's how it'd work. When casting, the upper bits are truncated - so, for example:
```rust
let x = 258_usize;
println!("x: {} -> {}", x, x as u8);
```
Prints out:
```
x: 258 -> 2
```
And not:
```
x: 258 -> 255
```

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
````diff
@@ -0,0 +1,677 @@
+    ///     sparse_path.at_depth(depth).unwrap(),
+    /// );
+    /// ```
+    pub fn at_idx(&self, index: usize) -> Option<RpoDigest> {
+        // If this overflows *or* if the depth is zero then the index was out of bounds.
+        let depth = NonZero::new(u8::checked_sub(self.depth(), index as u8)?)?;
+        self.at_depth(depth).ok()
+    }
````
Do we actually need this method? Seems like it is currently used only for tests and it is not clear to me why we'd want this to be a part of the public interface.

gswirski replied:
Removed from both `SparseMerklePath` and `MarklePath`.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,677 @@
+        (self.empty_nodes_mask & Self::bitmask_for_depth(node_depth)) != 0
+    }
+
+    fn get_nonempty_index(&self, node_depth: NonZero<u8>) -> usize {
```
Let's add doc comments here to explain what this function does. My understanding is that it takes node depth and returns the node's index in the `self.nodes` vector - correct? And if the node at this depth is empty, would it panic?

An alternative could be to combine this method with `is_depth_empty()` to look something like this:
```rust
fn get_node_index(&self, node_depth: NonZero<u8>) -> Option<usize> {
    ...
}
```
If the node at the specified depth is empty, it would just return `None`. I think this could lead a slightly cleaner code elsewhere.

gswirski replied:
Good catch! It wouldn't panic, it would return incorrect results. I've kept the name (I think it better represents the intent), but added a doc comment and changed the return type to `Option<size>`. It indeed makes the call site cleaner.

bobbinth commented on miden-crypto/src/merkle/sparse_path.rs:
```diff
@@ -0,0 +1,677 @@
+    // PROVIDERS
+    // ============================================================================================
+
+    /// Constructs a borrowing iterator over the nodes in this path.
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = RpoDigest> {
```
Could we specify the direction of iteration here? i.e., is it from the root to the leaf or from the leaf to the root?

gswirski replied:
Added a comment both here and on the `SparseMerklePathIter` (it's a public type).

bobbinth reviewed (commented):
Looks good! Thank you. I left some more comments inline.
